b2g/components/RemoteControlService.jsm

```diff
@@ -491,30 +624,342 @@ function getCommandJSSandbox(aImportFunctions) {
+      this._clientID = aEvent.detail.id;
+      let keys = this._server.getKeysFromUUID(aEvent.detail.id);
+      this._aes256B64 = { value: keys.aesKey };
+      this._hmac256B64 = { value: keys.hmacKey };
+    }
+
+    this._connection.sendMsg(reply);
+    this._status = EVENT_HANDLER_STATUS.ROUND_1;
+  },
+
+  _handleJPAKERound1Event: function(aEvent) {
+    if (this._status != EVENT_HANDLER_STATUS.ROUND_1) {
+      DEBUG && debug("Status not in ROUND_1, drop the event");
+      this._handleError(this, aEvent.type, "Wrong status");
+      return;
+    }
+
+    // First time connection, generate PIN, trigger notification on screen
+    if (this._clientID === null) {
+      let pin = this._server.getPIN();
+      if(pin === null) {
+        pin = this._server.generatePIN();
+        // Show notification on screen
+        SystemAppProxy._sendCustomEvent(REMOTE_CONTROL_EVENT, { pincode: pin, action: 'pin-created' });
+      }
+    }
+
+    // Store client round 1
+    this._client_round1 = aEvent.detail;
+
+    // Compute server round 1
+    this._JPAKE = Cc["@mozilla.org/services-crypto/sync-jpake;1"]
+                    .createInstance(Ci.nsISyncJPAKE);
+    try {
+      this._JPAKE.round1(TV_SIGNER_ID, this._gx1, this._gv1, this._r1,
+                         this._gx2, this._gv2, this._r2);
+    } catch (e) {
+      DEBUG && debug("JPAKE round1 get error: " + e);
+      this._handleError(this, aEvent.type, "JPAKE error: " + e.message);
+      return;
+    }
+
+    // Send round_1 result
+    let reply = {
+      type: "auth",
+      action: "jpake_server_1",
+      detail: {
+        gx1: this._gx1.value,
+        gx2: this._gx2.value,
+        zkp_x1: {gr: this._gv1.value, b: this._r1.value, id: TV_SIGNER_ID},
+        zkp_x2: {gr: this._gv2.value, b: this._r2.value, id: TV_SIGNER_ID}
+      }
+    };
+
+    this._connection.sendMsg(reply);
+    this._status = EVENT_HANDLER_STATUS.ROUND_2;
+  },
+
+  _handleJPAKERound2Event: function(aEvent) {
+    if (this._status != EVENT_HANDLER_STATUS.ROUND_2) {
+      DEBUG && debug("Status not in ROUND_2, drop the event");
+      this._handleError(this, aEvent.type, "Wrong status");
+      return;
+    }
+
+    // Store client round 2
+    this._client_round2 = aEvent.detail;
+
+    // Compute JPAKE round 2
+    let pin = null;
+
+    if (this._clientID !== null) {
+      // Not first time connection, use first 4 characters of previous AES key as PIN code
+      // Full AES key + TLS server cert is too long, round2 will throw error
+      pin = this._aes256B64.value.slice(0, PINCODE_LENGTH);
+    } else {
+      pin = this._server.getPIN();
+      if (pin === null) {
+        // Maybe some other client already used the PIN then clean it
+        DEBUG && debug("PN expire");
+        this._handleError(this, aEvent.type, "PIN expire");
+        return;
+      }
+    }
+
+    // Secret is PIN + first 12 characters of TLS server cert fingerprint
+    let secret = pin + this._server.getCertFingerprint();
```

Comment:
Using only 12 characters of the fingerprint is too little. Use the full fingerprint.
You should map the secret into the group, that way you won't get errors (the error probably stem from too big numbers or numbers that are not in the group.)